js/ui/components/paygManager.js

```diff
@@ -0,0 +1,154 @@
+// -*- mode: js; js-indent-level: 4; indent-tabs-mode: nil -*-
+//
+// Copyright (C) 2018 Endless Mobile, Inc.
+//
+// This is a GNOME Shell component to wrap the interactions over
+// D-Bus with the eos-payg system daemon.
+//
+// Licensed under the GNU General Public License Version 2
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+const Gio = imports.gi.Gio;
+const GLib = imports.gi.GLib;
+
+const Main = imports.ui.main;
+const Lang = imports.lang;
+const Signals = imports.signals;
+
+const EOS_PAYG_NAME = 'com.endlessm.Payg1';
+const EOS_PAYG_PATH = '/com/endlessm/Payg1';
+
+const EOS_PAYG_IFACE = '<node> \
+<interface name="com.endlessm.Payg1"> \
+<method name="AddCode"> \
+  <arg type="ay" direction="in" name="code"/> \
+</method> \
+<method name="ClearCode" /> \
+<signal name="Expired" /> \
+<property name="ExpiryTime" type="t" access="read"/> \
+<property name="Enabled" type="b" access="read"/> \
+</interface> \
+</node>';
+
+var PaygManager = new Lang.Class({
+    Name: 'PaygManager',
+
+    _init: function() {
+        this._proxy = null;
+        this._proxyInfo = Gio.DBusInterfaceInfo.new_for_xml(EOS_PAYG_IFACE);
+
+        this._paygEnabled = false;
+        this._expiryTime = 0;
+
+        this._codeExpiredId = 0;
+        this._propertiesChangedId = 0;
+    },
+
+    _onPropertiesChanged: function(proxy, changedProps, invalidatedProps) {
+        let propsDict = changedProps.deep_unpack();
+        if (propsDict.hasOwnProperty('Enabled'))
+            this._setEnabled(this._proxy.Enabled);
+
+        if (propsDict.hasOwnProperty('ExpiryTime'))
+            this._setExpiryTime(this._proxy.ExpiryTime);
+    },
+
+    _setEnabled: function(value) {
+        if (this._paygEnabled === value)
+            return;
+
+        this._paygEnabled = value;
+        this.emit('payg-enabled', this._paygEnabled);
+    },
+
+    _setExpiryTime: function(value) {
+        if (this._expiryTime === value)
+            return;
+
+        this._expiryTime = value;
+        this.emit('payg-expiry-time-changed', this._expiryTime);
```

Comment:
Not `notify::expiry-time`? I guess `expiry-time` isn’t actually a property on this class, is it?

Reply:
Hmm... I guess it could be `notify::expiry-time`, yes. Probably good to make `notify::enabled` too?

Reply:
Depends whether those are actually exposed as GObject properties. If not, I think it would be incorrect and confusing to use the `notify` signal.

Reply:
Well, that's a good point. In that case I think `enabled-changed` and `expiry-time-changed` makes more sense then, since PaygManager is not really a GObject, but a GJS class.

To be fair, I've also thought that would not be right if not emitting from a GObject, but the case of calendar.js made me hesitate since it emits a `notify::has-calendars` from `DBusEventSource` even if that's not a GObject either: https://github.com/endlessm/gnome-shell/blob/master/js/ui/calendar.js#L232

I'll change it then